Add support for Android controllers

britzl opened the pull request:
Fixes #3588

britzl committed:
Initial commit

britzl committed:
Update default.gamepads

britzl committed:
Update android_joystick.c

britzl commented:
Things remaining to be solved:

* Take into account axis deadzone
   * Described here: https://developer.android.com/training/game-controllers/controller-input#joystick
   * Need to figure out how to do this using the NDK
* Need to detect gamepad disconnects
   * Connects are done on the first button or axis event
   * Should use InputDevice.getDeviceIds()
   * https://developer.android.com/training/game-controllers/controller-input#input
* Detected gamepads are currently using the device id as index into the fixed array of possible connected gamepads (max 16). The device id is an int32_t so it is very likely to fall outside of the expected range. Need to create a lookup between a device id to gamepad index
* 
Check NDK samples: https://github.com/android/ndk-samples

JCash commented:
>Take into account axis deadzone

We do it ourselves in hid.cpp, and the dead zone is configurable via the `.gamepads` file.

britzl committed:
Added Android gamepad connect/disconnect

britzl committed:
Cleanup

britzl commented:
> * Take into account axis deadzone

Not required. Handled in input.cpp

> * Need to detect gamepad disconnects

Added detection in DefoldActivity.java

> * Detected gamepads are currently using the device id as index into the fixed array of possible connected gamepads

Added lookup between device id and index

britzl commented on engine/glfw/java/com/dynamo/android/DefoldActivity.java:
```diff
@@ -464,4 +469,46 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        }
+
+        // create and copy device ids to an int[]
+        int[] deviceIds = new int[mGameControllerDeviceIds.size()];
```
I'm not super happy with this allocation but I couldn't figure out another way to go from ArrayList<Integer> to int[]

britzl committed:
Indentation

JCash commented on engine/glfw/lib/android/android_joystick.c:
```diff
@@ -1,37 +1,459 @@
+            // Clear joystick state
+            for(n = 0; n < _glfwJoy[joystickIndex].NumAxes; ++n)
+            {
+                _glfwJoy[joystickIndex].Axis[n] = 0.0f;
```
Just setting something to 0, we often use "memset(array, 0, sizeof(array)"

JCash commented on engine/glfw/lib/android/android_joystick.c:
```diff
@@ -1,37 +1,459 @@
+            }
+            for(n = 0; n < _glfwJoy[joystickIndex].NumButtons; ++n)
+            {
+                _glfwJoy[joystickIndex].Button[n] = GLFW_RELEASE;
```
... which might or might not work well with such constants, but usually they're chosen to be default==false==0

JCash replied:
Code gets slightly shorter, but might still be nice since we want smaller executables, and the readability isn't really affected imho.

https://godbolt.org/z/3nxMY79j5

britzl replied:
Good point. GLFW_RELEASE is 0 so memsetting to 0 works.

JCash reviewed (dismissed):
Nice work!

britzl commented on engine/glfw/lib/android/android_joystick.c:
```diff
@@ -1,37 +1,459 @@
+            _glfwJoy[joystickIndex].DeviceId = deviceId;
+            _glfwJoy[joystickIndex].NumAxes = GLFW_ANDROID_GAMEPAD_NUMAXIS;
+            _glfwJoy[joystickIndex].NumButtons = GLFW_ANDROID_GAMEPAD_NUMBUTTONS;
+            _glfwJoy[joystickIndex].Axis = (float *)malloc(sizeof(float) * _glfwJoy[joystickIndex].NumAxes);
```
Good point. Not sure why I didn't do that!

britzl committed:
Review fixes
